addons/pos_self_order/models/res_config_settings.py

```diff
@@ -67,7 +67,7 @@ def _onchange_pos_self_order_pay_after(self):
         if self.pos_self_ordering_service_mode == 'counter' and self.pos_self_ordering_mode == 'mobile':
             self.pos_self_ordering_pay_after = "each"
 
-        if self.pos_self_ordering_pay_after == "each" and not self.module_pos_preparation_display:
+        if self.pos_self_ordering_mode != 'nothing' and self.pos_self_ordering_pay_after == "each" and not self.module_pos_preparation_display:
```

Comment:
Hello @gteboul, I'm ok with the fix but I think we can go further and maybe include both `'nothing'` and `'consultation'`. Both of them don't let you see the field `self_ordering_pay_after`. 

The commit you have mentioned (https://github.com/odoo/odoo/commit/8ee8010cf67f20c4eb28a334cc943bfedf2dd0d1)  mentioned that his issue was happening when orders are paid upfront, `'consultation'` mode (aka Qr Code only) won't give you the option to pay, it only lets you browse the menu.